---
title: "DESeq2 analysis for subset of human Ebola virus RNA-Seq dataset"
output: 
  html_document:
    keep_md: true
---
Transcript counts were obtained previously using featureCounts.

Load needed libraries and the data:
```{r message=FALSE, warning=FALSE, paged.print=TRUE}
library("DESeq2")
library("vsn")
library("RColorBrewer")
library("pheatmap")
library("EnhancedVolcano")

counts <- read.table('output/counts.txt', skip=1, header=TRUE, row.names="Geneid")
drops <- c("Chr","Start", "End", "Strand", "Length")
counts <- counts[, !(names(counts) %in% drops)]

head(counts[,c(1,2)])
```
Format the sample names nicely, and create a factor that groups the samples properly into convalescent, survival, and fatal outcomes:
```{r}
cols <- colnames(counts)
coldata <- matrix(nrow=dim(counts)[2],ncol=1)
n <- 1
for(x in cols){
   x <- substring(x,first = 53,last = nchar(x)-4)
   cols[n] <- x
   coldata[n] <- x
   n <- n+1
}
colnames(counts) <- cols
rownames(coldata) <- coldata[,1]
colnames(coldata) <- c("outcome")
coldata <- as.data.frame(coldata)
coldata$outcome <- substring(coldata$outcome,1,3)
coldata$outcome <- factor(coldata$outcome)

head(coldata)
```
Make the DESeq data object:
```{r}
dds <- DESeqDataSetFromMatrix(countData=counts,colData=coldata,design=~outcome)

dds
```
Discard very lowly expressed genes:
```{r}
keep <- rowSums(counts(dds)) >= 30
dds <- dds[keep,]

dds
```
Do the actual DE analysis, and look at summaries of the comparisons:
```{r message=FALSE, warning=FALSE}
dds <- DESeq(dds)
resFAT <- results(dds, name="outcome_FAT_vs_CON")
resSUR <- results(dds, name="outcome_SUR_vs_CON")

summary(resFAT, alpha=0.01)
summary(resSUR, alpha=0.01)
```
Do shrinkage of the LFCs:
```{r message=FALSE, warning=FALSE}
resFAT_LFCape <- lfcShrink(dds,coef="outcome_FAT_vs_CON", type="apeglm")
resFAT_LFCnorm <- lfcShrink(dds,coef="outcome_FAT_vs_CON", type="norm")

resSUR_LFCape <- lfcShrink(dds,coef="outcome_SUR_vs_CON", type="apeglm")
resSUR_LFCnorm <- lfcShrink(dds,coef="outcome_SUR_vs_CON", type="norm")
```
Compare the two LFC shrinkage methods:
```{r}
plotMA(resFAT,main='fatal vs. convalescent, original FCs',ylim=c(-10,10))
plotMA(resFAT_LFCape,main='fatal vs. convalescent, apeglm FCs',ylim=c(-10,10))
plotMA(resFAT_LFCnorm,main='fatal vs. convalescent, norm FCs',ylim=c(-10,10))
```
```{r}
plotMA(resSUR,main='survive vs. convalescent, original FCs',ylim=c(-4,4))
plotMA(resSUR_LFCape,main='survive vs. convalescent, apeglm FCs',ylim=c(-4,4))
plotMA(resSUR_LFCnorm,main='survive vs. convalescent, norm FCs',ylim=c(-4,4))
```

The normal shrinkage method appears to perform best for these data.

Sort the results by p-value and inspect them:
```{r}
resFAT_norm_ordered <- resFAT_LFCnorm[order(resFAT_LFCnorm$pvalue),]
resSUR_norm_ordered <- resSUR_LFCnorm[order(resSUR_LFCnorm$pvalue),]

head(resFAT_norm_ordered,3)
head(resSUR_norm_ordered,3)
```
Make volcano plots for the two comparisons:
```{r,fig.width=7, fig.height=10}
EnhancedVolcano(resFAT_LFCnorm, lab=rownames(resFAT_LFCnorm),x='log2FoldChange',y='pvalue',title='fatal vs. convalescent')
EnhancedVolcano(resSUR_LFCnorm, lab=rownames(resSUR_LFCnorm),x='log2FoldChange',y='pvalue',title='survive vs. convalescent')
```

Plot the dispersion estimates for quality control:
```{r}
plotDispEsts(dds)
```

Some visual analysis (heatmap, PCA) on a normalized version of the counts:
```{r,fig.height=5}
ntd <- normTransform(dds)

sampleDists <- dist(t(assay(ntd)))
sampleDistMatrix <- as.matrix(sampleDists)
rownames(sampleDistMatrix) <- ntd$outcome
colnames(sampleDistMatrix) <- ntd$outcome
colors <- colorRampPalette( rev(brewer.pal(9, "Blues")) )(255)
pheatmap(sampleDistMatrix,
         clustering_distance_rows=sampleDists,
         clustering_distance_cols=sampleDists,
         col=colors)
```

The outcomes are generally grouped together, with a few exceptions. Three of the fatal samples appear to be very "far" from most of the other samples.
```{r}
plotPCA(ntd,intgroup="outcome")
```

PCA appears to group the outcomes together decently well. Looks like there's more variation within the fatal and survival groups than the convalescent group, which agrees with the notion of entropy.

Finally, calculate the entropies for the two comparisons:
```{r}
# fatal vs. convalescent entropy
0.5*log(2*pi*exp(1)*var(resFAT_LFCnorm$log2FoldChange))

# survival vs. convalescent entropy
0.5*log(2*pi*exp(1)*var(resSUR_LFCnorm$log2FoldChange))
```